drivers/sensor/bq274xx/bq274xx.c

```diff
@@ -90,236 +101,293 @@ static int bq274xx_ctrl_reg_write(const struct device *dev, uint16_t subcommand)
 	return 0;
 }
 
-static int bq274xx_read_data_block(const struct device *dev, uint8_t offset,
-				   uint8_t *data, uint8_t bytes)
+static int bq274xx_get_device_type(const struct device *dev, uint16_t *val)
 {
-	const struct bq274xx_config *config = dev->config;
-	uint8_t i2c_data;
 	int ret;
 
-	i2c_data = BQ274XX_EXT_BLKDAT_START + offset;
-
-	ret = i2c_burst_read_dt(&config->i2c, i2c_data, data, bytes);
+	ret = bq274xx_ctrl_reg_write(dev, BQ274XX_CTRL_DEVICE_TYPE);
 	if (ret < 0) {
-		LOG_ERR("Failed to read block");
+		LOG_ERR("Unable to write control register");
 		return -EIO;
 	}
 
-	k_sleep(BQ274XX_SUBCLASS_DELAY);
+	ret = bq274xx_cmd_reg_read(dev, BQ274XX_CMD_CONTROL, val);
+	if (ret < 0) {
+		LOG_ERR("Unable to read register");
+		return -EIO;
+	}
 
 	return 0;
 }
 
-static int bq274xx_get_device_type(const struct device *dev, uint16_t *val)
+static int bq274xx_read_block(const struct device *dev,
+			      uint8_t subclass,
+			      uint8_t *block, uint8_t num_bytes)
 {
+	const struct bq274xx_config *const config = dev->config;
 	int ret;
 
-	ret = bq274xx_ctrl_reg_write(dev, BQ274XX_CTRL_DEVICE_TYPE);
+	ret = i2c_reg_write_byte_dt(&config->i2c, BQ274XX_EXT_DATA_CLASS, subclass);
 	if (ret < 0) {
-		LOG_ERR("Unable to write control register");
+		LOG_ERR("Failed to update state subclass");
 		return -EIO;
 	}
 
-	ret = bq274xx_cmd_reg_read(dev, BQ274XX_CMD_CONTROL_LOW, val);
+	/* DataBlock(), 0 for the first 32 bytes. */
+	ret = i2c_reg_write_byte_dt(&config->i2c, BQ274XX_EXT_DATA_BLOCK, 0x00);
 	if (ret < 0) {
-		LOG_ERR("Unable to read register");
+		LOG_ERR("Failed to update block offset");
+		return -EIO;
+	}
+
+	k_sleep(BQ274XX_SUBCLASS_DELAY);
+
+	ret = i2c_burst_read_dt(&config->i2c, BQ274XX_EXT_BLKDAT_START, block, num_bytes);
+	if (ret < 0) {
+		LOG_ERR("Unable to read block data");
 		return -EIO;
 	}
 
 	return 0;
 }
 
-static int bq274xx_gauge_configure(const struct device *dev)
+static int bq274xx_write_block(const struct device *dev,
+			       uint8_t *block, uint8_t num_bytes)
 {
 	const struct bq274xx_config *const config = dev->config;
-	struct bq274xx_data *data = dev->data;
-	const struct bq274xx_regs *regs = data->regs;
+	uint8_t checksum = 0;
 	int ret;
-	uint8_t tmp_checksum, checksum_old, checksum_new;
-	uint16_t flags, designenergy_mwh, taperrate, reg_val;
-	uint8_t block[BQ27XXX_DM_SZ];
-	uint8_t try;
 
-	designenergy_mwh = (uint32_t)config->design_capacity * 37 / 10; /* x3.7 */
-	taperrate = config->design_capacity * 10 / config->taper_current;
-
-	/* Unseal the battery control register */
-	ret = bq274xx_ctrl_reg_write(dev, BQ274XX_UNSEAL_KEY_A);
+	ret = i2c_burst_write_dt(&config->i2c, BQ274XX_EXT_BLKDAT_START, block, num_bytes);
 	if (ret < 0) {
-		LOG_ERR("Unable to unseal the battery");
+		LOG_ERR("Unable to read block data");
```

Comment:
Hi @fabiobaltieri unfortunately I've just tested my bq27441 device and it could not be initialised due to:

[00:00:01.344,787] <err> bq274xx: bq274xx_write_block: Unable to read block data

Also:
```suggestion
		LOG_ERR("Unable to write block data");
```

Reply:
Well that's unfortunate, would you be able to debug it? The Linux driver has an extra 1ms sleep before writing, maybe that would help? I'll fix the log message.

Reply:
I haven't been able to figure it out as yet :(

If I do a git bisect on the commits it's this one 🤔
`sensor: bq274xx: rewrite the configuration function`

Reply:
Yeah sure that's the one that introduced the block write, no big surprise there. But the Linux driver does that write as well so it gotta work with other parts too, the difference though is that the Linux driver does the page selection again and has a 1ms sleep so if that fixes it we can do it too.

BTW are you really using a 441 or is that a typo? This driver should only support the 421 and 427 right now.

Reply:
Yes we're using the 441 it just seems to 'work' atm.

Reply:
I'm shocked that TI went through the hassle of changing the device ID for the 427 (which apparently is the same silicon as the 426) and did not for the 441, I'm imagining it reports as 421. Anyway, whatever, let's sort this thing out.

Have you tried with the extra sleep or something? My best guess would be either that 1ms mystery sleep from the linux code, or writing the subclass and data block selection again.

I can add them to the PR if that makes it easier for you.